cli/tools/test/mod.rs

```diff
@@ -1020,6 +1028,127 @@ fn is_supported_test_ext(path: &Path) -> bool {
   }
 }
 
+struct TestFilterCollector<'a> {
+  should_run: bool,
+  filter: &'a str,
+}
+
+impl<'a> TestFilterCollector<'a> {
+  fn new(filter: &'a str) -> Self {
+    Self {
+      should_run: false,
+      filter,
+    }
+  }
+}
+
+impl Visit for TestFilterCollector<'_> {
+  fn visit_call_expr(&mut self, node: &ast::CallExpr) {
+    // Short circuit if we already know that we need to run this file
+    if self.should_run {
+      return;
+    }
+
+    if let ast::Callee::Expr(callee_expr) = &node.callee {
```

Comment:
What happens if someone doesn't use `Deno.test` (ex. bdd from deno_std) or mixes some other style with `Deno.test`?

Reply:
I don't account for that in the PR. I'm happy to add the other common test aliases like bdd and others. I'll gladly trade the speed boost for that. Other runners like `jest` or `vitest` require you to explicitly call `it/test/describe` because they do a very similar analysis as proposed in this PR.

Note that our lsp only detects `Deno.test` and doesn't understand the bdd API either.

Reply:
Yeah, I think we should handle it because otherwise it would be a regression in this case. Overall I think this is a good idea to do. You might be able to use the source files from `crate::cache::ParsedSourceCache` to make it even faster (those are conditionally parsed by deno_graph and only freed when loading the module, which would be before this).

Reply:
Good point, will add the `describe/it` family on Monday. Didn't know about `ParsedSourceCache`. That's a great tip 👍

Reply:
@nayeemrmn could we somehow handle `describe/it` in the LSP to run them in test explorer?

Reply:
> @nayeemrmn could we somehow handle `describe/it` in the LSP to run them in test explorer?

Just collect those in `TestCollector`, you'll get code lens and test explorer support.

I want to add that it's not just `describe()`/`it()` or well known testing libraries, users will sometimes have project/file specific test wrappers to do some setup or whatever. You won't be able to use `--filter` for those anymore. I don't think it's a blocker but if people complain we might later have to follow up with a way to opt out of this.

Reply:
Actually, let's make it so it runs the file anyway when it's a single file specified on the CLI. That can be the bypass for users that need one. Or maybe run any file which is explicitly named (not discovered or glob-matched) on the CLI regardless of this opt.

Reply:
We should probably also run a file if it has zero matches as the user might be using something else.

Btw, could we share the test collector code in the lsp with the code here? That would probably be most ideal.

Reply:
@nayeemrmn Looks like this is the issue for `it/describe` not being detected by our lsp: https://github.com/denoland/deno/issues/14304

@dsherret agree, that's a thought that crossed my mind too. The test discovery mechanism here is very similar to the one in the LSP. Seems like there is an opportunity to use the same abstraction.

I'm not sure if running a file when `Deno.test` is not present is worth it. I feel like we might overestimate the number of users who write their own abstractions around `Deno.test` or the bdd API. I can see though how this can be thought of as a breaking change. Maybe we can put it behind `--unstable` for now and then make it the default sometime later?

Reply:
> I'm not sure if running a file when Deno.test is not present is worth it. I feel like we might overestimate the number of users who write their own abstractions around Deno.test or the bdd API.

Not worth it for the engineering effort or the performance hit? For the engineering effort, once using the same code as the lsp I believe this would just be checking if anything was found or not (so very low effort) and for performance this wouldn't affect users who are using `Deno.test` or `bdd` because they would have test files that contain one of these.

Reply:
True, good point. Was thinking of the perf hit, but hadn't considered that this doesn't matter if all your test files contain `Deno.test`.